### PR TITLE
Persistent storage for append /caching performance

### PR DESCRIPTION
**Pull request by aphelionz**
This PR:
- Implements the persistent store work encapsulated in [`orbit-db-storage-adapter`](https://github.com/orbitdb/orbit-db-storage-adapter) and its dependents.
- Removes the `test/utils` folder, favoring the updated [`orbit-db-test-utils`](https://github.com/orbitdb/orbit-db-test-utils) package.
   - by proxy drops these dependencies:  `go-ipfs-dep`, `http-server`, `ipfs`, `is-node`, `ipfs-repo`, `ipfsd-ctl`, `is-node`, `istanbul`
- Restores the loading visualizer tool in the examples folder
- Fixes browser tests by removing the `--disable-dev-shm-usage` flag and forking [`mocha-headless-chrome`](https://github.com/aphelionz/mocha-headless-chrome)

### CODE REVIEW

File: src/entry-index.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+'use strict'
```
Comment (shamb0t) [inline src/entry-index.js]: Is this file necessary?

File: benchmarks/runner/benchmarks/append.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ const base = {
     return { log, ipfs, repo }
   },
   cycle: async function ({ log }) {
-    await log.append('Hello')
+    await log.append('Hello', 32)
```
Comment (shamb0t) [inline benchmarks/runner/benchmarks/append.js]: Why 32 here?

Comment (aphelionz) [reply benchmarks/runner/benchmarks/append.js]: In the `orbit-db-store` PR I changed the default `referenceCount` to 32, so I changed it here to make the benchmark more accurately reflect that.

Comment (shamb0t) [reply benchmarks/runner/benchmarks/append.js]: :+1:

File: test/log-join-concurrent.spec.js [hunk 3]
```diff
@@ -37,6 +37,8 @@ Object.keys(testAPIs).forEach(IPFS => {
 
     after(async () => {
       await stopIpfs(ipfs)
+      console.log(testIdentity)
```
Comment (shamb0t) [inline test/log-join-concurrent.spec.js]: Can probably remove this

File: examples/entry.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+const { MemStore } = require('orbit-db-test-utils')
+const IdentityProvider = require('orbit-db-identity-provider')
+
+module.exports = {
+  Log,
+  EntryIO,
+  Ipfs,
+  MemStore,
+  IdentityProvider
+}
```
Comment (haadcode) [inline examples/entry.js]: What's the purpose of this example? There's not a lot to it, so perhaps add a comment describing what to look for, or if this is not intended as an example, move the file to a more appropriate place.

File: .circleci/config.yml [hunk 5]
```diff
@@ -39,9 +39,9 @@ jobs:
     # 'See docs on artifact collection here https://circleci.com/docs/2.0/artifacts/'
     - run: mkdir -p $CIRCLE_ARTIFACTS $CIRCLE_TEST_REPORTS
     # This is based on your 1.0 configuration file or project settings
-    - run:
-        working_directory: ~/orbitdb/ipfs-log
-        command: nvm install 10.15.0 && nvm alias default 10.15.0
+    # - run:
+    #     working_directory: ~/orbitdb/ipfs-log
+    #     command: nvm install 10.15.0 && nvm alias default 10.15.0
```
Comment (haadcode) [inline .circleci/config.yml]: This is a small thing, but something that I think we should do: can we split this particular change into its own PR? It's fairly separate from the core of this PR and would be good to make explicit in the commit history (in master), that the CI config was changed.

File: test/log-join-concurrent.spec.js [hunk 6]
```diff
@@ -37,6 +37,8 @@ Object.keys(testAPIs).forEach(IPFS => {
...
+      await testIdentity._provider._keystore._store.close()
```
Comment (haadcode) [inline test/log-join-concurrent.spec.js]: This feels a bit weird (but admittedly I don't have the full context) as generally properties that are prefixed with `_` (underscore) are intended to be private (as far as private properties/variables go in JS). Is there a way to call `close()` "directly", eg. `testIdentity.provider.keystore.store.close()`? If not, what would it require to do so, instead of access the "private" variables directly?

File: test/signed-log.spec.js [hunk 7]
```diff
@@ -5,14 +5,19 @@ const rmrf = require('rimraf')
...
   testAPIs,
   startIpfs,
-  stopIpfs
-} = require('./utils')
+  stopIpfs,
+  implementations
+} = require('orbit-db-test-utils')
+
+const properLevelModule = implementations.filter(i => i.key.indexOf('level') > -1).map(i => i.module)[0]
+const storage = require('orbit-db-storage-adapter')(properLevelModule)
```
Comment (haadcode) [inline test/signed-log.spec.js]: Could these two lines be moved to test-utils? Then here, one would only need smth like `const { defaultStorage } = require('orbit-db-test-utils'); const storage = defaultStorage`. Doing so would remove the need to specify these two lines in each test.

File: test/signed-log.spec.js [hunk 8]
```diff
@@ -25,14 +30,22 @@ Object.keys(testAPIs).forEach((IPFS) => {
...
       rmrf.sync(signingKeysPath)
       await fs.copy(identityKeyFixtures, identityKeysPath)
       await fs.copy(signingKeyFixtures, signingKeysPath)
-      testIdentity = await IdentityProvider.createIdentity({ id: 'userA', identityKeysPath, signingKeysPath })
-      testIdentity2 = await IdentityProvider.createIdentity({ id: 'userB', identityKeysPath, signingKeysPath })
+
+      identityStore = await storage.createStore(identityKeysPath)
+      signingStore = await storage.createStore(signingKeysPath)
+      const keystore = new Keystore(identityStore)
+      const signingKeystore = new Keystore(signingStore)
```
Comment (haadcode) [inline test/signed-log.spec.js]: Could this functionality (creating the storage/store) be moved to IdentityProvider, so that it creates those by default? I feel this adds extra hoops for users who wish to use IdentityProvider to create an identity (note: one line) and making it as simple as possible (one line) for the user provides a better UX. What do you think?

File: test/entry-io.spec.js [hunk 9]
```diff
@@ -28,16 +33,30 @@ Object.keys(testAPIs).forEach((IPFS) => {
...
-      testIdentity4 = await IdentityProvider.createIdentity({ id: 'userD', identityKeysPath, signingKeysPath })
+      const defaultOptions = { identityKeysPath, signingKeysPath }
+
+      identityStore = await storage.createStore(identityKeysPath)
+      signingStore = await storage.createStore(signingKeysPath)
+      const keystore = new Keystore(identityStore)
+      const signingKeystore = new Keystore(signingStore)
+
+      const users = ['userA', 'userB', 'userC', 'userD']
+      options = await Promise.all(users.map(async (user) => {
```
Comment (haadcode) [inline test/entry-io.spec.js]: I  believe this doesn't need to be wrapped in `Promise.all` as the function to apply (Object.assign) is sync. Can prolly remove the Promise.all wrap?

File: test/replicate.spec.js [hunk 10]
```diff
@@ -110,6 +125,10 @@ Object.keys(testAPIs).forEach((IPFS) => {
         log2 = new Log(ipfs2, testIdentity2, { logId })
         input1 = new Log(ipfs1, testIdentity, { logId })
         input2 = new Log(ipfs2, testIdentity2, { logId })
+        const addr1 = await ipfs1.id()
+        const addr2 = await ipfs2.id()
+        await ipfs1.swarm.connect(addr2.addresses[0])
+        await ipfs2.swarm.connect(addr1.addresses[0])
```
Comment (haadcode) [inline test/replicate.spec.js]: We could use `connectPeers()` from https://github.com/orbitdb/orbit-db-test-utils/blob/master/connect-peers.js here to fold these four lines to one.

File: test/log-join-concurrent.spec.js (hunk 6 above)
Comment (shamb0t) [reply test/log-join-concurrent.spec.js]: We could either add a keystore getter to the provider or add a `close()` to the provider which calls `this._keystore.close()`, then this can be replaced with `testIdentity.provider.close()`, I would prefer adding a `close()` to the provider

File: examples/entry.js (hunk 4 above)
Comment (haadcode) [reply examples/entry.js]: This should still be remove, no?

Comment (aphelionz) [reply examples/entry.js]: Negative. This is the entry point for the webpack example build, so we need that to build the bundle which runs the examples

**Comment by haadcode** [comment]
LGTM 👍
